### DISCUSSION TITLE
Setup instructions and requirements

### DISCUSSION THREAD

**Discussion by cruftex** [envkey/envkey#107] [Ideas and feedback]
We are considering to use envkey. 

To get an idea of the effort to install a new software and what kind of server is required, I always take a look at the installation instructions. Unfortunately, I don't get any information, only run "install.sh" or make a file executable.

Some questions I have and hoped to get answers in the installation documentation:

- What are the system requirements? Minimum OS version? Memory? Storage?
- Is there a daemon running? How is it starting?
- How do I access the UI? Is it web based?
- How do clients access the server?
- Is it a client server architecture anyhow? What about an architecture diagram?
- How can I achieve redundancy?
- Where is the data stored? How do I back it up?
- How am I doing an update?
- How am I doing an uninstall?
- Where does the service log to?
- What do I need to do to secure the service?
- What and who is an "admin"?
- How is the password data stored? Is it some database engine we already use and know how to secure, debug and backup ?
- How is the audit data stored?
- Can the audit logs be forwarded to a central audit log server?

Although, it is convenient to just run a shell script, this is a no-go. Either standards need to be supported, like a package manager, or it should be documented what is the program directory, the data directory and how a system service managed and started, etc. How can an IDS check the integrity of the installation?

Why isn't gpg used for signing?

Maybe most of my questions would be answered if I just give it a try and play with it. But I don't run install.sh on any server I manage. That's a catch 22.

I know that all of this is a lot of work. So, I am writing in deep respect and maybe I am asking the questions that others have and don't bother asking.

**Comment by danenania** [comment]
Hi @cruftex, 

You should find answers to many of your questions, particularly on installing with verification (using minisign), architecture, and security in the docs, beginning with installation instructions: https://docs-v2.envkey.com/docs/install

Answers to the rest of your questions below are included inline:

What are the system requirements? Minimum OS version? Memory? Storage?

- EnvKey runs on all versions of Mac, Windows, and Linux, going back at least several years. Memory and storage depend on usage, but are in typical ranges for desktop-based software.

Is there a daemon running? How is it starting?

- Yes, it is managed by the EnvKey UI and CLI. There isn't a need to manage the daemon yourself.

How do I access the UI? Is it web based?

- It's an installable desktop app.

How do clients access the server?

- Via an API.

Is it a client server architecture anyhow? What about an architecture diagram?

- Yes, it's a client-server architecture. The architecture is described here: https://docs-v2.envkey.com/docs/components

How can I achieve redundancy?

- EnvKey Cloud and  Business Self-Hosted options include multiple levels of redundancy.

Where is the data stored? How do I back it up?

- In a MySQL database. You can export data from an org for backups via the UI or CLI.

How am I doing an update?

- Updates are automatic.

How am I doing an uninstall?

- Delete the app and CLI binary.

Where does the service log to?

- `$HOME/.envkey/logs`

What do I need to do to secure the service?

- It depends on how you're deploying. For EnvKey Cloud, you don't need to do anything. For Business Self-Hosted, there is a guide here: https://docs-v2.envkey.com/docs/enterprise-self-hosted For community self-hosted, it is up to you.

What and who is an "admin"?

- An Org Admin has access to all apps and can also invite additional users at an Org Admin level or below.

How is the password data stored? Is it some database engine we already use and know how to secure, debug and backup ?

- It is end-to-end encrypted and stored in a MySQL database.

How is the audit data stored?

- In a MySQL database.

Can the audit logs be forwarded to a central audit log server?

- Not currently.

**Comment by cruftex** [comment]
Thanks for the fast answers! 

You wrote:
> You should find answers to many of your questions, particularly on installing with verification (using minisign), architecture, and security in the docs, beginning with installation instructions: https://docs-v2.envkey.com/docs/install

That URL is the one that I was reading.

You wrote;
> - Updates are automatic.

That is also a no-go. How can I fulfil security requirements like controlled updates and documentation? A secure server should not be able to do outgoing connections to the internet, nor should software modify itself.

Please consider a more standards based approach in Unix/Linux environments, e.g. syslog, package based distribution and updates etc. I have the feeling that it is currently hard to impossible to integrate the software into an existing, standards based, enterprise environment.

**Reply by danenania** [reply]
@cruftex Sorry, I should have clarified that UI updates are automatic. When self-hosting with Business Self-Hosted, server-side updates are triggered by the user.

Also with Business Self-Hosted, servers runs behind a firewall and cannot make outgoing connections. Updates are done through AWS codebuild.

**Reply by danenania** [reply]
But overall, it sounds like given your requirements, EnvKey may not be the right solution for you. Good luck on finding a better fit.
